genesis/genesis.go

```diff
@@ -28,18 +28,174 @@ import (
+// *Config is not considered valid.
+func validateConfig(networkID uint32, config *Config) error {
+	if networkID != config.NetworkID {
+		return fmt.Errorf(
+			"networkID %d specified but genesis config contains networkID %d",
+			networkID,
+			config.NetworkID,
+		)
+	}
+
+	initialSupply, err := config.InitialSupply()
+	switch {
+	case err != nil:
+		return fmt.Errorf("unable to calculate initial supply: %w", err)
+	case initialSupply == 0:
+		return errors.New("initial supply must be > 0")
+	}
+
+	startTime := time.Unix(int64(config.StartTime), 0)
+	if time.Since(startTime) < 0 {
+		return fmt.Errorf(
+			"start time cannot be in the future: %s",
+			startTime,
+		)
+	}
+
+	// We don't impose any restrictions on the minimum
+	// stake duration to enable complex testing configurations
+	// but recommend setting a minimum duration of at least
+	// 15 minutes.
+	if config.InitialStakeDuration == 0 {
+		return errors.New("initial stake duration must be > 0")
+	}
+
+	if len(config.InitialStakers) == 0 {
+		return errors.New("initial stakers must be > 0")
+	}
+
+	offsetTimeRequired := config.InitialStakeDurationOffset * uint64(len(config.InitialStakers)-1)
+	if offsetTimeRequired > config.InitialStakeDuration {
+		return fmt.Errorf(
+			"initial stake duration is %d but need at least %d with offset of %d",
+			config.InitialStakeDuration,
+			offsetTimeRequired,
+			config.InitialStakeDurationOffset,
+		)
+	}
+
+	if err := validateInitialStakedFunds(config); err != nil {
+		return fmt.Errorf("initial staked funds validation failed: %w", err)
+	}
+
+	if len(config.CChainGenesis) == 0 {
+		return errors.New("C-Chain genesis cannot be empty")
+	}
+
+	if len(config.Message) == 0 {
+		return errors.New("genesis message cannot be empty")
+	}
+
+	return nil
+}
+
 // Genesis returns the genesis data of the Platform Chain.
 //
 // Since an Avalanche network has exactly one Platform Chain, and the Platform
 // Chain defines the genesis state of the network (who is staking, which chains
 // exist, etc.), defining the genesis state of the Platform Chain is the same as
 // defining the genesis state of the network.
 //
-// The ID of the new network is [networkID].
+// Geneses accepts:
+// 1) The ID of the new network. [networkID]
+// 2) The location of a custom genesis config to load. [filepath]
+//
+// If [filepath] is empty or the given network ID is Mainnet or Testnet, loads the
+// network genesis state from predefined configs. If [filepath] is non-empty and networkID
+// isn't Mainnet or Testnet, loads the network genesis data from the config at [filepath].
+//
+// Genesis returns:
+// 1) The byte representation of the genesis state of the platform chain
+//    (ie the genesis state of the network)
+// 2) The asset ID of AVAX
+func Genesis(networkID uint32, filepath string) ([]byte, ids.ID, error) {
+	config := GetConfig(networkID)
+	if len(filepath) > 0 && networkID != constants.MainnetID && networkID != constants.TestnetID {
```

Comment:
Could you change this to return an error if a genesis file path is specified despite specifying a pre-defined network? Perhaps: `cannot override genesis for predefined network`. I think we should probably include the local network ID here as well, since if that networkID is used then other libraries might assume that they know the correct constants to use for the AVAXAssetID, blockchainIDs, etc.